smac/facade/hydra_facade.py

```diff
@@ -130,20 +252,124 @@ def optimize(self):
-                self.solver = SMAC(scenario=self.scenario, tae_runner=self.tae, **self.kwargs)
         self.rh.save_json(fn=os.path.join(self.top_dir, 'all_runs_runhistory.json'), save_external=True)
         with open(os.path.join(self.top_dir, 'portfolio.pkl'), 'wb') as fh:
-            pickle.dump(portfolio, fh)
+            pickle.dump(self.portfolio, fh)
         self.logger.info("~"*120)
         self.logger.info('Resulting Portfolio:')
-        for configuration in portfolio:
+        for configuration in self.portfolio:
             self.logger.info(str(configuration))
         self.logger.info("~"*120)
 
-        return portfolio
+        return self.portfolio
+
+    def _update_portfolio(self, incs: np.ndarray) -> typing.Union[np.float, float]:
+        """
+        Validates all configurations (in incs) and determines which ones to add to the portfolio
+
+        Parameters
+        ----------
+        incs: np.ndarray
+            List of Configurations
+
+        """
+        self.logger.info('*'*120)
+        self.logger.info('Validating')
+        new_rh = self.solver.validate(config_mode=incs,
+                                      instance_mode=self.val_set,
+                                      repetitions=1,
+                                      use_epm=False,
+                                      n_jobs=self.n_optimizers)
+        self.rh.update(new_rh, origin=DataOrigin.EXTERNAL_SAME_INSTANCES)
+        self.logger.info("Number of validated runs: %d", (len(new_rh.data)))
+        self.logger.info('*'*120)
+        self.logger.info('Determining best incumbents')
+        results = []
+        config_cost_per_inst = {}
+        for incumbent in incs:
+            cost_per_inst = new_rh.get_instance_costs_for_config(config=incumbent)
+            config_cost_per_inst[incumbent] = cost_per_inst
+            results.append(np.mean(list(self.cost_per_inst.values())))
+        to_keep_ids = list(map(lambda x: x[0],
+                               sorted(enumerate(results), key=lambda y: y[1])))[:self.incs_per_round]
+        if len(to_keep_ids) > 1:
+            self.logger.info('Keeping incumbents of runs %s', ', '.join(map(str, to_keep_ids)))
+        else:
+            self.logger.info('Keeping incumbent of run %s', str(to_keep_ids))
+        keep_incumbents = incs[to_keep_ids]
+        for kept in keep_incumbents:
+            self.portfolio.append(kept)
+            cost_per_inst = config_cost_per_inst[kept]
+            if self.cost_per_inst:
+                if len(self.cost_per_inst) != len(cost_per_inst):
+                    raise ValueError('Num validated Instances mismatch!')
+                for key in cost_per_inst:
+                    self.cost_per_inst[key] = min(self.cost_per_inst[key], cost_per_inst[key])
+            else:
+                self.cost_per_inst = cost_per_inst
+
+        cur_cost = np.mean(list(self.cost_per_inst.values()))  # type: np.float
+        return cur_cost
+
+    def _get_rng(
```

Comment:
looks like a function that should not live in Hydra exclusively.

Reply:
SMAC also uses this _get_rng function.
Should they both inherit it from a super class?